### DISCUSSION TITLE
Compilation flows in practice

### DISCUSSION THREAD

**Discussion by csccva** [AdaptiveCpp/AdaptiveCpp#1050] [General]
Hello, 

I am little confused about this. How I enable one or the other? 
I compiled OPenSYCL with support for  `omp.accelerated` and support for the MI250X. Are there some flags when I compile my code to select one or the other?
Also which one of these `hip.integrated-multipass` or `hip.explicit-multipass` ma I using and how does it affect performance?

Cristian

**Comment by illuhad** [comment]
Hi!

you can always explicitly select a compilation flow that you want to use: `--hipsycl-targets="omp.accelerated"` / `--hipsycl-targets="omp.library-only"`  / `--hipsycl-targets="hip.integrated-multipass:gfx906"` / ...

If you just use the `omp`, `hip` or `cuda` abbreviations, it will select one of the flavors as default choice based on some rules that are described here:
https://github.com/OpenSYCL/OpenSYCL/blob/develop/doc/using-hipsycl.md#abbreviations
(again, if you want to depart from the defaults, just pass the full name of the compilation flow you wish to use).

If you just use `--hipsycl-targets=hip:gfxXYZ`, you are using integrated multipass.
The difference between integrated and explicit multipass is described here: https://github.com/OpenSYCL/OpenSYCL/blob/develop/doc/compilation.md

Kernel performance between the two should be exactly the same, as the same device compiler is used. The only difference is that for explicit multipass, `syclcc` controls the device compiler invocation, while in integrated multipass it lets clang handle both host and device compilation.
Explicit multipass is mainly needed when you want to target both CUDA and HIP devices from a single binary while still using an SMCP model (instead of the new generic single-pass compiler). In practice with explicit multipass the main difference you might feel compared to integrated multipass is that interoperability guarantees with CUDA/HIP source code are slightly different in the host pass: https://github.com/OpenSYCL/OpenSYCL/blob/develop/doc/compilation.md#language-extension-guarantees

**Reply by csccva** [reply]
Thank you for the  detailed explanation. Things are a little clearer now.
